crates/re_renderer/examples/picking.rs

```diff
@@ -0,0 +1,205 @@
+use itertools::Itertools as _;
+use rand::Rng;
+use re_renderer::{
+    view_builder::{Projection, TargetConfiguration, ViewBuilder},
+    Color32, PickingLayerId, PickingLayerInstanceId, PointCloudBuilder, RenderContext,
+    ScheduledPickingRect, Size,
+};
+
+mod framework;
+
+struct Picking {
+    points_positions: Vec<glam::Vec3>,
+    points_radii: Vec<Size>,
+    points_colors: Vec<Color32>,
+    points_picking_ids: Vec<PickingLayerInstanceId>,
+
+    scheduled_picking_rects: Vec<ScheduledPickingRect>,
+
+    picking_position: glam::UVec2,
+}
+
+fn random_color(rnd: &mut impl rand::Rng) -> Color32 {
+    ecolor::Hsva {
+        h: rnd.gen::<f32>(),
+        s: rnd.gen::<f32>() * 0.5 + 0.5,
+        v: rnd.gen::<f32>() * 0.5 + 0.5,
+        a: 1.0,
+    }
+    .into()
+}
+
+impl Picking {
+    #[allow(clippy::unused_self)]
+    fn handle_incoming_picking_data(&mut self, re_ctx: &mut RenderContext, _time: f32) {
+        re_ctx
+            .gpu_readback_belt
+            .lock()
+            .receive_data(|data, identifier| {
+                if let Some(index) = self
+                    .scheduled_picking_rects
+                    .iter()
+                    .position(|s| s.identifier == identifier)
+                {
+                    let picking_rect_info = self.scheduled_picking_rects.swap_remove(index);
+
+                    // TODO(andreas): Move this into a utility function?
+                    let picking_data_without_padding =
+                        picking_rect_info.row_info.remove_padding(data);
+                    let picking_data: &[PickingLayerId] =
+                        bytemuck::cast_slice(&picking_data_without_padding);
+
+                    // Grab the middle pixel. usually we'd want to do something clever that snaps the the closest object of interest.
+                    let picked_pixel = picking_data[(picking_rect_info.extent.x / 2
+                        + (picking_rect_info.extent.y / 2) * picking_rect_info.extent.x)
+                        as usize];
+                    if picked_pixel.object.0 != 0 {
+                        let index = picked_pixel.instance.0
+                            + (self.points_radii.len() as u64 / 2) * (picked_pixel.object.0 - 1);
```

Comment:
wat

this makes no sense to me, having read the code so far. Why no store the point index directly in the picking buffer?

If you want/need complicated logic for mapping points indices to `PickingLayerId`, then wrap it up into two symmetric functions next to each other so the logic is made clear

Reply:
yeah, this is a huge hack...